sources/iTermApplicationDelegate.m

```diff
@@ -501,18 +502,19 @@ - (BOOL)application:(NSApplication *)theApplication openFile:(NSString *)filenam
         // Verify whether filename is a script or a folder
         BOOL isDir;
         [[NSFileManager defaultManager] fileExistsAtPath:filename isDirectory:&isDir];
-        if (!isDir) {
-            NSString *aString = [NSString stringWithFormat:@"%@; exit;\n", [filename stringWithEscapedShellCharacters]];
-            [[iTermController sharedInstance] launchBookmark:nil inTerminal:[self terminalToOpenFileIn]];
-            // Sleeping a while waiting for the login.
-            sleep(1);
-            [[[[iTermController sharedInstance] currentTerminal] currentSession] insertText:aString];
+        iTermController *controller = [iTermController sharedInstance];
+
+        if (isDir) {
+            NSMutableDictionary *bookmark = [[[controller defaultBookmark] mutableCopy] autorelease];
+            bookmark[KEY_WORKING_DIRECTORY] = filename;
+            bookmark[KEY_CUSTOM_DIRECTORY] = kProfilePreferenceInitialDirectoryCustomValue;
+            [controller launchBookmark:bookmark inTerminal:[self terminalToOpenFileIn]];
         } else {
-            NSString *aString = [NSString stringWithFormat:@"cd %@\n", [filename stringWithEscapedShellCharacters]];
-            [[iTermController sharedInstance] launchBookmark:nil inTerminal:[self terminalToOpenFileIn]];
-            // Sleeping a while waiting for the login.
+            [controller launchBookmark:nil inTerminal:[self terminalToOpenFileIn]];
             sleep(1);
```

Comment:
You could get rid of this sleep by creating a `bookmark` like above and doing this:
````
if ([bookmark[KEY_INITIAL_TEXT] length] > 0) {
  bookmark[KEY_INITIAL_TEXT] = [bookmark[KEY_INITIAL_TEXT] stringByAppendingFormat:@"\n%@; exit", [filename stringWithEscapedShellCharacters]]];
} else {
  bookmark[KEY_INITIAL_TEXT] = [NSString stringWithFormat:@"%@; exit", [filename StringWIthEscapedShellCharacters]];
}